Makefile: add environment variable to specify Binary Ninja path

sweishen opened the pull request:
Closes #65

I notice that `install_api.py` isn't bundled. Should we actually update to use `$BINARY_NINJA_PATH/scripts/install_api.py` instead of `./install_api.py` by default?

sweishen commented on Makefile.venv:
```diff
@@ -10,10 +11,16 @@ $(VENV_PATH):
 	virtualenv -ppython$(PYTHON_VERSION) $(VENV_PATH)
 	. $(VENV_PATH)/bin/activate
 	pip install -r $(PREFIX)/requirements.txt
-	pip install -r $(PREFIX)/requirements-compiler-idioms.txt
```
Fixed, I'm not sure why I committed this in the first place.

sweishen commented on Makefile.venv:
```diff
@@ -10,10 +11,16 @@ $(VENV_PATH):
+$(VENV_PATH)/lib/python$(PYTHON_VERSION)/site-packages/binaryninja.pth: $(BINARY_NINJA_PATH)/python
 ifneq ($(wildcard $(PREFIX)/install_api.py),"")
 	. $(VENV_PATH)/bin/activate && \
 		$(VENV_PATH)/bin/python $$(readlink -f $(PREFIX)/install_api.py) && \
```
Updated.

mari-mari commented:
So it works like a charm with my default settings - Binary Ninja in `/opt/binaryninja`.

However, when I move Binary Ninja and try to run like offered, I get the following error:
```
 ➜ make BINARY_NINJA_PATH=/opt/bkp/binaryninja
make -f Makefile.venv venv
make[1]: Entering directory '/home/mry/.binaryninja/plugins/dewolf'
make[1]: Nothing to be done for 'venv'.
make[1]: Leaving directory '/home/mry/.binaryninja/plugins/dewolf'
. .venv/bin/activate
PYTHONPATH=. pytest --ignore-glob="*test-lifting.py" --ignore-glob="tests/test_sample_binaries.py" --ignore-glob="tests/test_plugin.py" tests
ImportError while loading conftest '/home/mry/.binaryninja/plugins/dewolf/conftest.py'.
__init__.py:7: in <module>
    from binaryninja import BackgroundTaskThread, BinaryView, Function, MessageBoxButtonSet, PluginCommand, core_ui_enabled, show_html_report
E   ModuleNotFoundError: No module named 'binaryninja'
make: *** [Makefile:95: unittests] Error 4
```
Apparently it silently fails to install Binja in this case. Any ideas what could be the reason?

sweishen commented:
> Apparently it silently fails to install Binja in this case. Any ideas what could be the reason?

Looks like I missed out one instance of replacing /opt/binaryninja. Thanks for pointing it out.

On a separate note, I get this:

```
ERROR: Could not open requirements file: [Errno 2] No such file or directory: './requirements-compiler-idioms.txt'
```

Apparently there is no `requirements-compiler-idioms.txt` on the `main` branch of this repo. Perhaps we should make a copy from somewhere?

mari-mari commented:
Hm, I do not see `reply` option :\

> Apparently there is no `requirements-compiler-idioms.txt` on the `main` branch of this repo. Perhaps we should make a copy from somewhere?

Yep, it expects that you have `dewolf-idioms` in `./binaryninja/plugins` folder. It is the case in the VM, since the `Dockerfile` does the job of copying the requirements and renaming them. However, it is still not the case when running `make` locally.
We have a separate issue for that (https://github.com/fkie-cad/dewolf/issues/64)

If you insist, we could try to accomplish it in this pull request, but I generally prefer a separate issue since this one is ready for merging.

I still have a couple of points to this pull request though.

So I run `make`, it does not find Binary Ninja and prints an error message. Two things here:

- the virtual environment is being activated and I end up in Python shell.
```
Failed to find Binary Ninja at '/opt/binaryninja'.
Please set the BINARY_NINJA_PATH variable manually, e.g.
make ... BINARY_NINJA_PATH=/your/path/to/binja

. .venv/bin/activate && \
	.venv/bin/python $(readlink -f /opt/binaryninja/scripts/install_api.py) && \
	echo "/opt/binaryninja/python" > .venv/lib/python3.8/site-packages/binaryninja.pth
Python 3.8.10 (default, Mar 15 2022, 12:22:08) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```

Not critical, but maybe we could avoid activating virtual environment in case we are not able to install Binary Ninja?

- so I saw this error and I run `make BINARY_NINJA_PATH=...` with the right path. But since virtual environment already exists, `make` uses existing one, where Binary Ninja is not installed - since currently installation happens only when virtual environment is being created. This lets me with the following output:
```
make -f Makefile.venv venv
make[1]: Entering directory '/home/mry/.binaryninja/plugins/dewolf'
make[1]: Nothing to be done for 'venv'.
make[1]: Leaving directory '/home/mry/.binaryninja/plugins/dewolf'
. .venv/bin/activate
PYTHONPATH=. pytest --ignore-glob="*test-lifting.py" --ignore-glob="tests/test_sample_binaries.py" --ignore-glob="tests/test_plugin.py" tests
ImportError while loading conftest '/home/mry/.binaryninja/plugins/dewolf/conftest.py'.
__init__.py:7: in <module>
    from binaryninja import BackgroundTaskThread, BinaryView, Function, MessageBoxButtonSet, PluginCommand, core_ui_enabled, show_html_report
E   ModuleNotFoundError: No module named 'binaryninja'
make: *** [Makefile:95: unittests] Error 4
```
Running `make clean` before running `make BINARY_NINJA_PATH=...` does the job. But I wonder if we need to communicate to the user that `make clean` is needed beforehand. Or maybe add running `clean` target directly in case we ended with this error. That would remove `.venv` and will automatically solve both problems. 

So, both points are not critical. If you would like to address any of those, it would be great. If no - well, then I consider the pull request ready.

sweishen commented:
> Not critical, but maybe we could avoid activating virtual environment in case we are not able to install Binary Ninja?

This is really odd, but I've fixed it by using the virtualenv's pip directly.



> Running `make clean` before running `make BINARY_NINJA_PATH=...` does the job. But I wonder if we need to communicate to the user that `make clean` is needed beforehand. Or maybe add running `clean` target directly in case we ended with this error. That would remove `.venv` and will automatically solve both problems.

I've added another dependency so that this should get picked up correctly. Basically we were incorrectly writing `binaryninja.pth` even if we had not found `BINARY_NINJA_PATH`. It shouldn't be necessary to run `make clean` at all now.

mari-mari commented:
It seems to work locally but unfortunately the recent changes make the pipeline fail. Will have a look if I find a reason of that tomorrow.

sweishen commented:
I think the problem started when I tried use reference `.venv/bin/python` directly, but I haven't been able to figure it out. It also works locally for me. Will try to fix this when I have a chance.

sweishen commented:
```
 RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.8'
```

When building on my machine, I see

```
created virtual environment CPython3.8.10.final.0-64 in 150ms
```

I believe it's because we're pointing at ubuntu:latest, which was updated last week to 22.04, which has python3.10 by default. We should decide if we want to move to 3.10 or not, probably in another PR.

mari-mari commented:
> I believe it's because we're pointing at ubuntu:latest, which was updated last week to 22.04, which has python3.10 by default. We should decide if we want to move to 3.10 or not, probably in another PR.

Since moving to python3.10 did not worked trivially by changing `3.8` to `3.10` in `Makefile.venv`, we pinned Ubuntu 20.04 as a temporary solution, to have a running pipeline. And yes, I agree, we need a separate merge request where we will update to python3.10.
